gusto/time_discretisation.py

```diff
@@ -286,6 +284,24 @@ def setup(self, equation, apply_bcs=True, *active_labels):
         self.x0 = Function(self.fs)
         self.x1 = Function(self.fs)
 
+    @cached_property
+    def lhs(self):
+        """Set up the discretisation's left hand side (the time derivative)."""
+        l = self.residual.label_map(
+            lambda t: t.has_label(time_derivative),
+            map_if_true=replace_subject(self.x_out, self.idx),
+            map_if_false=drop)
+
+        return l.form
+
+    @cached_property
+    def solver(self):
+        """Set up the problem and the solver."""
+        # setup linear solver using lhs and rhs defined in derived class
+        problem = NonlinearVariationalProblem(self.lhs - self.rhs, self.x_out, bcs=self.bcs)
+        solver_name = self.field_name+self.__class__.__name__
+        return NonlinearVariationalSolver(problem, solver_parameters={'snes_type': 'ksponly'} | self.solver_parameters, options_prefix=solver_name)
```

Comment:
The `|` here is cool python to combine the dictionaries, so I'm glad to have learnt something new!

I wonder if you want to add a comment in the line above to explain why we need this and to explain the cool python. e.g:
```
# If snes_type not specified by user, set this to ksp only to avoid outer Newton iteration
```